base/broadcast.jl

```diff
@@ -3,63 +3,207 @@
             nullable_returntype, null_safe_op, hasvalue, isoperator
 import Base: broadcast, broadcast!
 export broadcast_getindex, broadcast_setindex!, dotview, @__dot__
 
-const ScalarType = Union{Type{Any}, Type{Nullable}}
+# Note: `indices` will be overridden below, thus you need to use
+# Base.indices when you want the Base versions.
+
+## Types used by `rule`
+# Unknown acts a bit like `Bottom`, in that it loses to everything. But by not having
+# it be a subtype of every other type, we limit the need for ambiguity resolution.
+abstract type Unknown end
+# Objects that act like a scalar for purposes of broadcasting
+abstract type Scalar end
+# An AbstractArray type that "loses" in precedence comparisons to all other AbstractArrays.
+# We will want to keep track of dimensionality, so we make it the first parameter.
+abstract type BottomArray{N} <: AbstractArray{Void,N} end
+Bottom0d     = BottomArray{0}
+BottomVector = BottomArray{1}
+BottomMatrix = BottomArray{2}
+# When two or more AbstractArrays have specialized broadcasting, and no `rule`
+# is defined to establish precedence, then we have a conflict
+abstract type ArrayConflict <: AbstractArray{Void,-1} end
+
+
+"""
+    result = Broadcast.Result{ContainerType}()
+    result = Broadcast.Result{ContainerType,ElType}(inds::Indices)
+
+Create an object that specifies the type and (optionally) indices
+of the result (output) of a broadcasting operation.
+
+Using a dedicated type for this information makes it possible to support
+variants of [`broadcast`](@ref) that accept `result` as an argument;
+it prevents an ambiguity of intent that would otherwise arise because
+both types and indices-tuples are among the supported *input*
+arguments to `broadcast`. For example, `parse.(Int, ("1", "2"))` is
+equivalent to `broadcast(parse, Int, ("1", "2"))`, and as a consequence
+it would would be ambiguous if result-type and output-indices information
+were passed as positional arguments to `broadcast`.
+
+You can extract `inds` with `indices(result)`.
+"""
+struct Result{ContainerType,ElType,I<:Union{Void,Indices}}
+    indices::I
+end
+Result{ContainerType}() where ContainerType =
+    Result{ContainerType,Void,Void}(nothing)
+Result{ContainerType,ElType}(inds::Indices) where {ContainerType,ElType} =
+    Result{ContainerType,ElType,typeof(inds)}(inds)
+indices(r::Result) = r.indices
+Base.indices(r::Result) = indices(r)
+Base.eltype(r::Result{ContainerType,ElType}) where {ContainerType,ElType} = ElType
+
+
+### User-extensible methods (see the Interfaces chapter of the manual) ###
+## Computing the result (output) type
+"""
+    Broadcast.rule(::Type{<:MyContainer}) = MyContainer
+
+Declare that objects of type `MyContainer` have a customized broadcast implementation.
+If you define this method, you are responsible for defining the following method:
+
+    Base.similar(f, r::Broadcast.Result{MyContainer}, As...) = ...
```

Comment:
This brings up an interesting point (apologies for the length here). While developing this, I found that subtyping caused some problems. Hence in the docs I recommend against it. Notice that for `rule` it uses `<:MyContainer` for the *input* but it deliberately returns the union-type `MyContainer` (with no parameters) as the *output*. It's the output that will be passed to `similar`, so this method definition is correct as stated.

But now I suspect that I encountered those problems back when I was using `Bottom` (`Union{}`) instead of defining new types, `Scalar` and `Unknown`. `Bottom` is a subtype of *any* type, and so it caused method ambiguities. Having switched to `Scalar`, I suspect it might be safe now to drop that recommendation (discussed further below).

From a purely conceptual standpoint, if unary `rule` preserved all the parameters of the input, one thing that would mildly annoy me is that `broadcast(+, MyArray([1,2,3]), 2.2)` would yield something like `Result{MyArray{Int,1}, Float64}`. The thing I don't like about this is that the `Float64` (the output eltype computed by broadcasting) seems to contradict the `Int` (which came from the input container). This is why in `BottomArray` I gave it an eltype of `Void`, but we can't automatically strip type parameters of arbitrary `AbstractArray` types. This is really more of a stylistic/purist issue, though.

Another issue I belatedly considered is that there might be some merit to recommending that the result of unary `rule` should be a concrete type, even if you have to make one up (`struct MyContainerWithNoEltype end`): [inference is reluctant to inline](https://github.com/JuliaLang/julia/blob/16139d054e29ea63b797537927beee84da1b229d/base/inference.jl#L17-L19) when types are not concrete.

That's a much longer answer than you were probably expecting! How much of this kind of detail should I go into?

Reply:
These details are great! Thank you! :)

Reply:
Thanks for the details. So indeed in this case it seems `<:` isn't needed. I was just concerned about using it for input types (IIRC for `Base.linearindices` we used to have examples which only worked when the type had no parameters).